Strange behaviors with cmake-wheels

While trying to migrate my plain `setup.py` to use the great `sckikit-build` tools, I encountered multiple problems. Some of then are reproducible.

## Problem 1: Requiring cmake on `pyproject.toml` does not seem to have any effect

### Dockerfile
```docker
FROM ubuntu:22.04
RUN apt-get update && apt-get install -y \
    python3-pip \
    build-essential \
    && rm -rf /var/lib/apt/lists/*
```

### Commands
```sh
git clone https://github.com/scikit-build/scikit-build-sample-projects.git
cd scikit-build-sample-projects/projects/hello-pybind11
pip install --verbose .
```

### Output

<details>

```sh
Using pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
Defaulting to user installation because normal site-packages is not writeable
Processing /workspace
  Running command pip subprocess to install build dependencies
  WARNING: The directory '/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
  Collecting setuptools
    Downloading setuptools-65.4.0-py3-none-any.whl (1.2 MB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 3.9 MB/s eta 0:00:00
  Collecting scikit-build>=0.13
    Downloading scikit_build-0.15.0-py2.py3-none-any.whl (77 kB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.5/77.5 KB 2.6 MB/s eta 0:00:00
  Collecting cmake
    Downloading cmake-3.24.1.1-py2.py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (23.3 MB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.3/23.3 MB 3.8 MB/s eta 0:00:00
  Collecting ninja
    Downloading ninja-1.10.2.3-py2.py3-none-manylinux_2_5_x86_64.manylinux1_x86_64.whl (108 kB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.1/108.1 KB 4.0 MB/s eta 0:00:00
  Collecting distro
    Downloading distro-1.7.0-py3-none-any.whl (20 kB)
  Collecting wheel>=0.29.0
    Downloading wheel-0.37.1-py2.py3-none-any.whl (35 kB)
  Collecting packaging
    Downloading packaging-21.3-py3-none-any.whl (40 kB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 40.8/40.8 KB 8.8 MB/s eta 0:00:00
  Collecting pyparsing!=3.0.5,>=2.0.2
    Downloading pyparsing-3.0.9-py3-none-any.whl (98 kB)
       ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.3/98.3 KB 3.2 MB/s eta 0:00:00
  Installing collected packages: ninja, cmake, wheel, setuptools, pyparsing, distro, packaging, scikit-build
  Successfully installed cmake-3.24.1.1 distro-1.7.0 ninja-1.10.2.3 packaging-21.3 pyparsing-3.0.9 scikit-build-0.15.0 setuptools-65.4.0 wheel-0.37.1
  Installing build dependencies ... done
  Running command Getting requirements to build wheel
  running egg_info
  creating src/hello_pybind11.egg-info
  writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
  writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
  Getting requirements to build wheel ... done
  Running command Preparing metadata (pyproject.toml)
  running dist_info
  creating /tmp/pip-modern-metadata-2b1jv5x8/hello_pybind11.egg-info
  writing manifest file '/tmp/pip-modern-metadata-2b1jv5x8/hello_pybind11.egg-info/SOURCES.txt'
  writing manifest file '/tmp/pip-modern-metadata-2b1jv5x8/hello_pybind11.egg-info/SOURCES.txt'
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: hello-pybind11
  Running command Building wheel for hello-pybind11 (pyproject.toml)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/setuptools_wrap.py", line 613, in setup
      cmkr = cmaker.CMaker(cmake_executable)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/cmaker.py", line 141, in __init__
      self.cmake_version = get_cmake_version(self.cmake_executable)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/cmaker.py", line 95, in get_cmake_version
      raise SKBuildError(
  Traceback (most recent call last):

  Problem with the CMake installation, aborting build. CMake executable is cmake
  error: subprocess-exited-with-error

  × Building wheel for hello-pybind11 (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> See above for output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /usr/bin/python3 /usr/lib/python3/dist-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmpbthvzkrp
  cwd: /workspace
  Building wheel for hello-pybind11 (pyproject.toml) ... error
  ERROR: Failed building wheel for hello-pybind11
Failed to build hello-pybind11
ERROR: Could not build wheels for hello-pybind11, which is required to install pyproject.toml-based projects
```

</details>

### Output Extract
```sh
  Running command Building wheel for hello-pybind11 (pyproject.toml)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/setuptools_wrap.py", line 613, in setup
      cmkr = cmaker.CMaker(cmake_executable)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/cmaker.py", line 141, in __init__
      self.cmake_version = get_cmake_version(self.cmake_executable)
    File "/tmp/pip-build-env-n_bmk7fx/overlay/local/lib/python3.10/dist-packages/skbuild/cmaker.py", line 95, in get_cmake_version
      raise SKBuildError(
  Traceback (most recent call last):
```

### Expected

build should work

### Workaround

`pip install cmake` will install the cmake-python-package globally, make the binary `$ cmake --version` available, and thus, the build will now work, but that comes to my second problem


## Problem 2, ninja and cmake runs at different times

After `pip install cmake` I can `pip install --verbose .` again the example and now it build, but somwthing strange is happening:
<details>

```sh
Using pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
Processing /workspace
  Running command pip subprocess to install build dependencies
  Collecting setuptools
    Using cached setuptools-65.4.0-py3-none-any.whl (1.2 MB)
  Collecting scikit-build>=0.13
    Using cached scikit_build-0.15.0-py2.py3-none-any.whl (77 kB)
  Collecting cmake
    Using cached cmake-3.24.1.1-py2.py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (23.3 MB)
  Collecting ninja
    Using cached ninja-1.10.2.3-py2.py3-none-manylinux_2_5_x86_64.manylinux1_x86_64.whl (108 kB)
  Collecting packaging
    Using cached packaging-21.3-py3-none-any.whl (40 kB)
  Collecting distro
    Using cached distro-1.7.0-py3-none-any.whl (20 kB)
  Collecting wheel>=0.29.0
    Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
  Collecting pyparsing!=3.0.5,>=2.0.2
    Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)
  Installing collected packages: ninja, cmake, wheel, setuptools, pyparsing, distro, packaging, scikit-build
  Successfully installed cmake-3.24.1.1 distro-1.7.0 ninja-1.10.2.3 packaging-21.3 pyparsing-3.0.9 scikit-build-0.15.0 setuptools-65.4.0 wheel-0.37.1
  WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
  Installing build dependencies ... done
  Running command Getting requirements to build wheel
  running egg_info
  writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
  Getting requirements to build wheel ... done
  Running command Preparing metadata (pyproject.toml)
  running dist_info
  creating /tmp/pip-modern-metadata-l04tfssq/hello_pybind11.egg-info
  writing manifest file '/tmp/pip-modern-metadata-l04tfssq/hello_pybind11.egg-info/SOURCES.txt'
  writing manifest file '/tmp/pip-modern-metadata-l04tfssq/hello_pybind11.egg-info/SOURCES.txt'
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: hello-pybind11
  Running command Building wheel for hello-pybind11 (pyproject.toml)
  Not searching for unused variables given on the command line.
  -- The C compiler identification is GNU 11.2.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- The CXX compiler identification is GNU 11.2.0
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /workspace/_cmake_test_compile/build
  -- The C compiler identification is GNU 11.2.0
  -- The CXX compiler identification is GNU 11.2.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  CMake Warning (dev) at /tmp/pip-build-env-ph_r9a33/overlay/local/lib/python3.10/dist-packages/cmake/data/share/cmake-3.24/Modules/FetchContent.cmake:1264 (message):
    The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
    not set.  The policy's OLD behavior will be used.  When using a URL
    download, the timestamps of extracted files should preferably be that of
    the time of extraction, otherwise code that depends on the extracted
    contents might not be rebuilt if the URL changes.  The OLD behavior
    preserves the timestamps from the archive instead, but this is usually not
    what you want.  Update your project to the NEW behavior or specify the
    DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
    robustness issue.
  Call Stack (most recent call first):
    CMakeLists.txt:9 (FetchContent_Declare)
  This warning is for project developers.  Use -Wno-dev to suppress it.

  -- pybind11 v2.9.2
  -- Found PythonInterp: /usr/bin/python3 (found version "3.10.6")
  -- Found PythonLibs: /usr/lib/x86_64-linux-gnu/libpython3.10.so
  -- Performing Test HAS_FLTO
  -- Performing Test HAS_FLTO - Success
  -- Configuring done
  -- Generating done
  CMake Warning:
    Manually-specified variables were not used by the project:

      Python3_EXECUTABLE
      Python3_INCLUDE_DIR
      Python3_LIBRARY
      Python_EXECUTABLE
      Python_INCLUDE_DIR
      Python_LIBRARY


  -- Build files have been written to: /workspace/_skbuild/linux-x86_64-3.10/cmake-build
  [1/3] Building CXX object CMakeFiles/_hello.dir/src/hello/hello_py.cpp.o
  [2/3] Linking CXX shared module _hello.cpython-310-x86_64-linux-gnu.so
  [2/3] Install the project...
  -- Install configuration: "Release"
  -- Installing: /workspace/_skbuild/linux-x86_64-3.10/cmake-install/src/hello/./_hello.cpython-310-x86_64-linux-gnu.so


  --------------------------------------------------------------------------------
  -- Trying "Ninja" generator
  --------------------------------
  ---------------------------
  ----------------------
  -----------------
  ------------
  -------
  --
  --
  -------
  ------------
  -----------------
  ----------------------
  ---------------------------
  --------------------------------
  -- Trying "Ninja" generator - success
  --------------------------------------------------------------------------------

  Configuring Project
    Working directory:
      /workspace/_skbuild/linux-x86_64-3.10/cmake-build
    Command:
      cmake /workspace -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/workspace/_skbuild/linux-x86_64-3.10/cmake-install/src/hello -DPYTHON_VERSION_STRING:STRING=3.10.6 -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/pip-build-env-ph_r9a33/overlay/local/lib/python3.10/dist-packages/skbuild/resources/cmake -DPython3_EXECUTABLE:FILEPATH=/usr/bin/python3 -DPython3_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPython3_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DPython_EXECUTABLE:FILEPATH=/usr/bin/python3 -DPython_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPython_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DPYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3 -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPYTHON_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DCMAKE_MAKE_PROGRAM:FILEPATH=/tmp/pip-build-env-ph_r9a33/overlay/local/lib/python3.10/dist-packages/ninja/data/bin/ninja -DCMAKE_BUILD_TYPE:STRING=Release


  copying src/hello/__init__.py -> _skbuild/linux-x86_64-3.10/cmake-install/src/hello/__init__.py

  running bdist_wheel
  running build
  running build_py
  creating _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10
  creating _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10/hello
  copying _skbuild/linux-x86_64-3.10/cmake-install/src/hello/__init__.py -> _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10/hello
  copying _skbuild/linux-x86_64-3.10/cmake-install/src/hello/_hello.cpython-310-x86_64-linux-gnu.so -> _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10/hello
  running build_ext
  running install
  running install_lib
  creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64
  creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel
  creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
  copying _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10/hello/_hello.cpython-310-x86_64-linux-gnu.so -> _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
  copying _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-3.10/hello/__init__.py -> _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
  running install_egg_info
  running egg_info
  writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
  Copying src/hello_pybind11.egg-info to _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello_pybind11-1.2.3.egg-info
  running install_scripts
  Building wheel for hello-pybind11 (pyproject.toml) ... done
  Created wheel for hello-pybind11: filename=hello_pybind11-1.2.3-cp310-cp310-linux_x86_64.whl size=47397 sha256=1b5a101225acabad0e455302820239eb22b02d0e497a28facf867c0a93eb7c1f
  Stored in directory: /tmp/pip-ephem-wheel-cache-m4kll0go/wheels/80/75/f6/d63789e3414f7f88206102fa596a8ddafc889f48984e8712b3
Successfully built hello-pybind11
Installing collected packages: hello-pybind11
Successfully installed hello-pybind11-1.2.3
```
</details>

### Extracted output

If you see, this part is quite strange to me
```sh
  --------------------------------------------------------------------------------
  -- Trying "Ninja" generator
  --------------------------------
  ---------------------------
  ----------------------
  -----------------
  ------------
  -------
  --
  --
  -------
  ------------
  -----------------
  ----------------------
  ---------------------------
  --------------------------------
  -- Trying "Ninja" generator - success
  --------------------------------------------------------------------------------
```

Looks like "ninja" was running late into the game.

## Problem 3, running everything inside a virtual env works, but it is not desired

So, after install `python3-venv` and install `pipx` I run this command `pipx run build --wheel .` inside the example project and get a more "excpected" output

### Output

<details>

```sh
* Creating venv isolated environment...
* Installing packages in isolated environment... (cmake, ninja, scikit-build>=0.13, setuptools)
* Getting dependencies for wheel...
running egg_info
creating src/hello_pybind11.egg-info
writing src/hello_pybind11.egg-info/PKG-INFO
writing dependency_links to src/hello_pybind11.egg-info/dependency_links.txt
writing top-level names to src/hello_pybind11.egg-info/top_level.txt
writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
reading manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
* Installing packages in isolated environment... (wheel)
* Building wheel...


--------------------------------------------------------------------------------
-- Trying "Ninja" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is GNU 11.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /workspace/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja" generator - success
--------------------------------------------------------------------------------

Configuring Project
  Working directory:
    /workspace/_skbuild/linux-x86_64-3.10/cmake-build
  Command:
    cmake /workspace -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/workspace/_skbuild/linux-x86_64-3.10/cmake-install/src/hello -DPYTHON_VERSION_STRING:STRING=3.10.6 -DSKBUILD:INTERNAL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/build-env-mv_bf13e/lib/python3.10/site-packages/skbuild/resources/cmake -DPython3_EXECUTABLE:FILEPATH=/tmp/build-env-mv_bf13e/bin/python -DPython3_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPython3_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DPython_EXECUTABLE:FILEPATH=/tmp/build-env-mv_bf13e/bin/python -DPython_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPython_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DPYTHON_EXECUTABLE:FILEPATH=/tmp/build-env-mv_bf13e/bin/python -DPYTHON_INCLUDE_DIR:PATH=/usr/include/python3.10 -DPYTHON_LIBRARY:PATH=/usr/lib/x86_64-linux-gnu/libpython3.10.so -DCMAKE_MAKE_PROGRAM:FILEPATH=/tmp/build-env-mv_bf13e/lib/python3.10/site-packages/ninja/data/bin/ninja -DCMAKE_BUILD_TYPE:STRING=Release

-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at /tmp/build-env-mv_bf13e/lib/python3.10/site-packages/cmake/data/share/cmake-3.24/Modules/FetchContent.cmake:1264 (message):
  The DOWNLOAD_EXTRACT_TIMESTAMP option was not given and policy CMP0135 is
  not set.  The policy's OLD behavior will be used.  When using a URL
  download, the timestamps of extracted files should preferably be that of
  the time of extraction, otherwise code that depends on the extracted
  contents might not be rebuilt if the URL changes.  The OLD behavior
  preserves the timestamps from the archive instead, but this is usually not
  what you want.  Update your project to the NEW behavior or specify the
  DOWNLOAD_EXTRACT_TIMESTAMP option with a value of true to avoid this
  robustness issue.
Call Stack (most recent call first):
  CMakeLists.txt:9 (FetchContent_Declare)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- pybind11 v2.9.2
-- Found PythonInterp: /tmp/build-env-mv_bf13e/bin/python (found version "3.10.6")
-- Found PythonLibs: /usr/lib/x86_64-linux-gnu/libpython3.10.so
-- Performing Test HAS_FLTO
-- Performing Test HAS_FLTO - Success
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    Python3_EXECUTABLE
    Python3_INCLUDE_DIR
    Python3_LIBRARY
    Python_EXECUTABLE
    Python_INCLUDE_DIR
    Python_LIBRARY


-- Build files have been written to: /workspace/_skbuild/linux-x86_64-3.10/cmake-build
[2/3] Install the project...
-- Install configuration: "Release"
-- Installing: /workspace/_skbuild/linux-x86_64-3.10/cmake-install/src/hello/./_hello.cpython-310-x86_64-linux-gnu.so

copying src/hello/__init__.py -> _skbuild/linux-x86_64-3.10/cmake-install/src/hello/__init__.py

running bdist_wheel
running build
running build_py
creating _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310
creating _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310/hello
copying _skbuild/linux-x86_64-3.10/cmake-install/src/hello/__init__.py -> _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310/hello
copying _skbuild/linux-x86_64-3.10/cmake-install/src/hello/_hello.cpython-310-x86_64-linux-gnu.so -> _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310/hello
copied 1 files
running build_ext
installing to _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel
running install
running install_lib
creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64
creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel
creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
copying _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310/hello/_hello.cpython-310-x86_64-linux-gnu.so -> _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
copying _skbuild/linux-x86_64-3.10/setuptools/lib.linux-x86_64-cpython-310/hello/__init__.py -> _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello
copied 2 files
running install_egg_info
running egg_info
writing src/hello_pybind11.egg-info/PKG-INFO
writing dependency_links to src/hello_pybind11.egg-info/dependency_links.txt
writing top-level names to src/hello_pybind11.egg-info/top_level.txt
reading manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
writing manifest file 'src/hello_pybind11.egg-info/SOURCES.txt'
Copying src/hello_pybind11.egg-info to _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello_pybind11-1.2.3-py3.10.egg-info
running install_scripts
copied 0 files
creating _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel/hello_pybind11-1.2.3.dist-info/WHEEL
creating '/workspace/dist/tmptmyz1qtm/hello_pybind11-1.2.3-cp310-cp310-linux_x86_64.whl' and adding '_skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel' to it
adding 'hello/__init__.py'
adding 'hello/_hello.cpython-310-x86_64-linux-gnu.so'
adding 'hello_pybind11-1.2.3.dist-info/METADATA'
adding 'hello_pybind11-1.2.3.dist-info/WHEEL'
adding 'hello_pybind11-1.2.3.dist-info/top_level.txt'
adding 'hello_pybind11-1.2.3.dist-info/RECORD'
removing _skbuild/linux-x86_64-3.10/setuptools/bdist.linux-x86_64/wheel
Successfully built hello_pybind11-1.2.3-cp310-cp310-linux_x86_64.whl
```

</details>

### Extract

The most interesting part is that now ninja is being found before cmake, and looks like the build is being run after cmake suceed. By the waye ** With this, I do not need to pip install cmake beforehand**

```sh

--------------------------------------------------------------------------------
-- Trying "Ninja" generator
--------------------------------
---------------------------
----------------------
-----------------
------------
-------
--
Not searching for unused variables given on the command line.
-- The C compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is GNU 11.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /workspace/_cmake_test_compile/build
--
-------
------------
-----------------
----------------------
---------------------------
--------------------------------
-- Trying "Ninja" generator - success
--------------------------------------------------------------------------------
```

## Problem 4: This creats problems with local installation of cmake

The fact that this is not working breaks also a local installation of cmake. For example, to remove the annoying warning about the timestamps (see all logs before) you can add this to your cmake:

```cmake
cmake_policy(SET CMP0135 OLD)
```

But, **withouth running pipx in a virtual enviroment** I get the following error on my dev machine (`cmake 3.22`):

```sh
  CMake Error at 3rdparty/find_dependencies.cmake:9 (cmake_policy):
    Policy "CMP0135" is not known to this version of CMake.
```

## Question

What can I do to help debugging this issue? Thanks in advance

--- a/skbuild/constants.py
+++ b/skbuild/constants.py
@@ -2,13 +2,31 @@
 This module defines constants commonly used in scikit-build.
 """
 
+import contextlib
 import os
 import platform
+import shutil
 import sys
 
 from distutils.util import get_platform
 
-CMAKE_DEFAULT_EXECUTABLE = "cmake"
+
+def _get_cmake_executable() -> str:
+    with contextlib.suppress(ModuleNotFoundError):
+        import cmake  # pylint: disable=import-outside-toplevel
+
+        return f"{cmake.CMAKE_BIN_DIR}/cmake"
+
+    for name in ("cmake3", "cmake"):
+        prog = shutil.which(name)
+        if prog:
+            return prog
+
+    # Just guess otherwise
+    return "cmake"
+
+
+CMAKE_DEFAULT_EXECUTABLE = _get_cmake_executable()
 """Default path to CMake executable."""
 
 

--- a/skbuild/constants.py
+++ b/skbuild/constants.py
@@ -7,6 +7,7 @@ import os
 import platform
 import shutil
 import sys
+from pathlib import Path
 
 from distutils.util import get_platform
 
@@ -15,7 +16,10 @@ def _get_cmake_executable() -> str:
     with contextlib.suppress(ModuleNotFoundError):
         import cmake  # pylint: disable=import-outside-toplevel
 
-        return f"{cmake.CMAKE_BIN_DIR}/cmake"
+        path = f"{cmake.CMAKE_BIN_DIR}/cmake"
+        if Path(f"{path}.exe").is_file():
+            return f"{path}.exe"
+        return path
 
     for name in ("cmake3", "cmake"):
         prog = shutil.which(name)

--- a/skbuild/setuptools_wrap.py
+++ b/skbuild/setuptools_wrap.py
@@ -588,21 +588,22 @@ def setup(  # noqa: C901
             dist = upstream_Distribution({"setup_requires": setup_requires})
             dist.fetch_build_eggs(setup_requires)  # type: ignore[no-untyped-call]
 
-            # Considering packages associated with "setup_requires" keyword are
-            # installed in .eggs subdirectory without honoring setuptools "console_scripts"
-            # entry_points and without settings the expected executable permissions, we are
-            # taking care of it below.
-            import cmake  # pylint: disable=import-outside-toplevel
+            with contextlib.suppress(ModuleNotFoundError):
+                # Considering packages associated with "setup_requires" keyword are
+                # installed in .eggs subdirectory without honoring setuptools "console_scripts"
+                # entry_points and without settings the expected executable permissions, we are
+                # taking care of it below.
+                import cmake  # pylint: disable=import-outside-toplevel
 
-            for executable in ("cmake", "cpack", "ctest"):
-                executable = os.path.join(cmake.CMAKE_BIN_DIR, executable)
-                if platform.system().lower() == "windows":
-                    executable += ".exe"
-                st = os.stat(executable)
-                permissions = st.st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
-                os.chmod(executable, permissions)
-            cmake_executable = os.path.join(cmake.CMAKE_BIN_DIR, "cmake")
-            break
+                for executable in ("cmake", "cpack", "ctest"):
+                    executable = os.path.join(cmake.CMAKE_BIN_DIR, executable)
+                    if platform.system().lower() == "windows":
+                        executable += ".exe"
+                    st = os.stat(executable)
+                    permissions = st.st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
+                    os.chmod(executable, permissions)
+                cmake_executable = os.path.join(cmake.CMAKE_BIN_DIR, "cmake")
+                break
 
     # Languages are used to determine a working generator
 

